app/api/items/ask_hint.go

```diff
@@ -0,0 +1,244 @@
+package items
+
+import (
+	"bytes"
+	"crypto/rsa"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"unsafe"
+
+	"github.com/go-chi/render"
+	"github.com/jinzhu/gorm"
+	"gopkg.in/jose.v1/crypto"
+
+	"github.com/France-ioi/AlgoreaBackend/app/database"
+	"github.com/France-ioi/AlgoreaBackend/app/logging"
+	"github.com/France-ioi/AlgoreaBackend/app/payloads"
+	"github.com/France-ioi/AlgoreaBackend/app/service"
+	"github.com/France-ioi/AlgoreaBackend/app/token"
+)
+
+func (srv *Service) askHint(w http.ResponseWriter, r *http.Request) service.APIError {
+	requestData := AskHintRequest{store: srv.Store, publicKey: srv.TokenConfig.PublicKey}
+
+	var err error
+	if err = render.Bind(r, &requestData); err != nil {
+		return service.ErrInvalidRequest(err)
+	}
+
+	user := srv.GetUser(r)
+	apiError := service.NoError
+	if apiError = checkAskHintUsers(user, &requestData); apiError != service.NoError {
+		return apiError
+	}
+
+	err = srv.Store.InTransaction(func(store *database.DataStore) error {
+		var hasAccess bool
+		var reason error
+		hasAccess, reason, err = store.Items().CheckSubmissionRights(requestData.TaskToken.Converted.LocalItemID, user)
+		service.MustNotBeError(err)
+
+		if !hasAccess {
+			apiError = service.ErrForbidden(reason)
+			return nil // commit! (CheckSubmissionRights() changes the DB sometimes)
+		}
+
+		userItemStore := store.UserItems()
+		err = userItemStore.CreateIfMissing(user.UserID, requestData.TaskToken.Converted.LocalItemID)
+		service.MustNotBeError(err)
+
+		// Get the previous hints requested JSON data
+		var hintsRequestedParsed []payloads.Anything
+		hintsRequestedParsed, err = queryAndParsePreviouslyRequestedHints(&requestData, store, user, r)
+		if err == gorm.ErrRecordNotFound {
+			apiError = service.ErrNotFound(errors.New("can't find previously requested hints info"))
+			return nil // commit
+		}
+		service.MustNotBeError(err)
+
+		// Add the new requested hint to the list if it's not in the list yet
+		hintsRequestedParsed = addHintToListIfNeeded(hintsRequestedParsed, requestData.HintToken.AskedHint)
+
+		var hintsRequestedNew []byte
+		hintsRequestedNew, err = json.Marshal(hintsRequestedParsed)
+		service.MustNotBeError(err)
+		hintsRequestedNewString := string(hintsRequestedNew)
+		requestData.TaskToken.HintsRequested = &hintsRequestedNewString
+		hintsGivenCountString := strconv.Itoa(len(hintsRequestedParsed))
+		requestData.TaskToken.HintsGivenCount = &hintsGivenCountString
+
+		columnsToUpdate := map[string]interface{}{
+			"sHintsRequested":            hintsRequestedNew,
+			"nbHintsCached":              len(hintsRequestedParsed),
+			"nbTasksWithHelp":            1,
```

Comment:
From the description "For how many of this item's descendants tasks within this attempts, did the user ask for help on the forum?", it does not seem relevant here and setting the value to "1" either (why not +1). I assume I copied that from PHP but I think it does not really make sense.

Reply:
Yes, it's from here: https://github.com/France-ioi/AlgoreaPlatform/blame/master/task/task.php#L286 . I'll ask Michel once he's available.

Reply:
> Michel [8:47 PM]
No, because it's, for any item, the number of tasks, *children included*, for which we requested help. Tasks don't have children, so it's either 0 at the start, or 1 once we request a hint ; on chapters, an user can't request hints, so this field gets calculated by the computeAllUserItems listener